InteractiveHtmlBom/ecad/kicad.py

```diff
@@ -548,7 +573,19 @@ def parse(self):
                 pcbdata["zones"] = {'F': [], 'B': []}
         if self.config.include_nets and hasattr(self.board, "GetNetInfo"):
             pcbdata["nets"] = self.parse_netlist(self.board.GetNetInfo())
-        components = [self.footprint_to_component(f) for f in self.footprints]
+
+        if self.config.extra_fields:
+            for f in pcbdata['footprints']:
+                if f['ref'] not in extra_field_data:
+                    # Some components are on pcb but not in schematic data.
+                    # Show a warning about possibly outdated netlist/xml file.
+                    self.logger.warn(
+                        'Component %s is missing from schematic data.' % f.ref)
+
+        components = [self.footprint_to_component(f,
+                      {} if extra_field_data is None
+                      else extra_field_data.get(f.GetReference()))
+                      for f in self.footprints]
```

Comment:
This list comprehension is overly complex. Split it into several statements for readability.
Also pass a default value to extra_field_data.get() for the case when reference is not found.